File: api/routes/damage_scenarios.py
"""
Damage Scenario API routes
"""
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
import logging

from api.deps.db import get_db
from api.models.damage_scenario import (
    DamageScenario, 
    DamageScenarioCreate, 
    DamageScenarioUpdate, 
    DamageScenarioList,
    PropagationSuggestion,
    PropagationSuggestionResponse
)
from api.services.damage_scenario_service import (
    create_damage_scenario as service_create_scenario,
    get_damage_scenario as service_get_scenario,
    get_damage_scenarios as service_get_scenarios,
    count_damage_scenarios as service_count_scenarios,
    update_damage_scenario as service_update_scenario,
    delete_damage_scenario as service_delete_scenario,
    generate_propagation_suggestions as service_generate_suggestions
)

router = APIRouter()
logger = logging.getLogger(__name__)


@router.get("", response_model=DamageScenarioList)
async def list_damage_scenarios(
    skip: int = 0, 
    limit: int = 100,
    scope_id: Optional[str] = None,
    component_id: Optional[str] = None,
    damage_category: Optional[str] = None,
    severity: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """
    List all damage scenarios with pagination and filtering
    """
    scenarios = service_get_scenarios(
        db, 
        skip=skip, 
        limit=limit,
        scope_id=scope_id,
        component_id=component_id,
        damage_category=damage_category,
        severity=severity
    )
    total = service_count_scenarios(
        db,
        scope_id=scope_id,
        component_id=component_id,
        damage_category=damage_category,
        severity=severity
    )
    return DamageScenarioList(scenarios=scenarios, total=total)


@router.post("", response_model=DamageScenario, status_code=status.HTTP_201_CREATED)
async def create_damage_scenario(
    scenario: DamageScenarioCreate,
    db: Session = Depends(get_db)
):
    """
    Create a new damage scenario
    """
    try:
        return service_create_scenario(db, scenario)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/{scenario_id}", response_model=DamageScenario)
async def get_damage_scenario(
    scenario_id: str, 
    db: Session = Depends(get_db)
):
    """
    Get a damage scenario by ID
    """
    scenario = service_get_scenario(db, scenario_id)
    if not scenario:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Damage scenario with ID {scenario_id} not found"
        )
    return scenario


@router.put("/{scenario_id}", response_model=DamageScenario)
async def update_damage_scenario(
    scenario_id: str, 
    scenario: DamageScenarioUpdate,
    db: Session = Depends(get_db)
):
    """
    Update a damage scenario
    """
    try:
        updated = service_update_scenario(db, scenario_id, scenario)
        if not updated:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Damage scenario with ID {scenario_id} not found"
            )
        return updated
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.delete("/{scenario_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_damage_scenario(
    scenario_id: str, 
    db: Session = Depends(get_db)
):
    """
    Delete a damage scenario
    """
    success = service_delete_scenario(db, scenario_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Damage scenario with ID {scenario_id} not found"
        )
    return None


@router.get("/propagation/suggestions", response_model=PropagationSuggestionResponse)
async def get_propagation_suggestions(
    component_id: str,
    confidentiality_impact: bool = Query(False, description="Whether confidentiality is impacted"),
    integrity_impact: bool = Query(False, description="Whether integrity is impacted"),
    availability_impact: bool = Query(False, description="Whether availability is impacted"),
    db: Session = Depends(get_db)
):
    """
    Get impact propagation suggestions for a component
    """
    try:
        suggestions = service_generate_suggestions(
            db, 
            component_id,
            confidentiality_impact,
            integrity_impact,
            availability_impact
        )
        return PropagationSuggestionResponse(suggestions=suggestions, total=len(suggestions))
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )

File: frontend/src/api/damage-scenarios.ts
/**
 * Damage Scenarios API client
 * Handles all API calls related to damage scenario management
 */

import apiClient from './index';

/**
 * Damage category enum
 */
export enum DamageCategory {
  PHYSICAL = "Physical",
  OPERATIONAL = "Operational",
  FINANCIAL = "Financial",
  PRIVACY = "Privacy",
  SAFETY = "Safety",
  ENVIRONMENTAL = "Environmental",
  REPUTATIONAL = "Reputational",
  LEGAL = "Legal",
  OTHER = "Other"
}

/**
 * Impact type enum
 */
export enum ImpactType {
  DIRECT = "Direct",
  INDIRECT = "Indirect",
  CASCADING = "Cascading"
}

/**
 * Severity level enum
 */
export enum SeverityLevel {
  LOW = "Low",
  MEDIUM = "Medium",
  HIGH = "High",
  CRITICAL = "Critical"
}

/**
 * Damage Scenario interface
 */
export interface DamageScenario {
  scenario_id: string;
  name: string;
  description: string;
  damage_category: DamageCategory;
  impact_type: ImpactType;
  confidentiality_impact: boolean;
  integrity_impact: boolean;
  availability_impact: boolean;
  severity: SeverityLevel;
  impact_details?: Record<string, any>;
  scope_id: string;
  primary_component_id: string;
  affected_component_ids: string[];
  version: number;
  revision_notes?: string;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * Damage Scenario list response
 */
export interface DamageScenarioList {
  scenarios: DamageScenario[];
  total: number;
}

/**
 * New damage scenario data (for creation)
 */
export type DamageScenarioCreate = Omit<DamageScenario, 'scenario_id' | 'is_deleted' | 'created_at' | 'updated_at'> & {
  scenario_id?: string;
};

/**
 * Damage scenario update data
 */
export type DamageScenarioUpdate = Partial<Omit<DamageScenario, 'scenario_id' | 'is_deleted' | 'created_at' | 'updated_at'>>;

/**
 * Propagation suggestion interface
 */
export interface PropagationSuggestion {
  source_component_id: string;
  target_component_id: string;
  confidentiality_impact: boolean;
  integrity_impact: boolean;
  availability_impact: boolean;
  damage_category: DamageCategory;
  impact_type: ImpactType;
  severity: SeverityLevel;
  confidence: number;
  path: string[];
}

/**
 * Propagation suggestion response
 */
export interface PropagationSuggestionResponse {
  suggestions: PropagationSuggestion[];
  total: number;
}

/**
 * Damage Scenarios API service
 */
export const damageScenarioApi = {
  /**
   * Get all damage scenarios with optional filters
   * @param options - Filter options
   * @returns Promise with array of damage scenarios
   */
  async getAll(options?: {
    skip?: number;
    limit?: number;
    scope_id?: string;
    component_id?: string;
    damage_category?: string;
    severity?: string;
  }): Promise<DamageScenarioList> {
    try {
      // Build query parameters
      const queryParams = new URLSearchParams();
      if (options?.skip !== undefined) queryParams.append('skip', options.skip.toString());
      if (options?.limit !== undefined) queryParams.append('limit', options.limit.toString());
      if (options?.scope_id) queryParams.append('scope_id', options.scope_id);
      if (options?.component_id) queryParams.append('component_id', options.component_id);
      if (options?.damage_category) queryParams.append('damage_category', options.damage_category);
      if (options?.severity) queryParams.append('severity', options.severity);
      
      const queryString = queryParams.toString();
      const endpoint = `/damage-scenarios${queryString ? `?${queryString}` : ''}`;
      
      const response = await apiClient.get<DamageScenarioList>(endpoint);
      return response;
    } catch (error) {
      console.error('Error fetching damage scenarios:', error);
      return { scenarios: [], total: 0 };
    }
  },

  /**
   * Get a damage scenario by ID
   * @param id - Damage scenario ID
   * @returns Promise with damage scenario data
   */
  async getById(id: string): Promise<DamageScenario> {
    return apiClient.get<DamageScenario>(`/damage-scenarios/${id}`);
  },

  /**
   * Create a new damage scenario
   * @param scenario - Damage scenario data
   * @returns Promise with created damage scenario
   */
  async create(scenario: DamageScenarioCreate): Promise<DamageScenario> {
    try {
      const response = await apiClient.post<DamageScenario>('/damage-scenarios', scenario);
      return response;
    } catch (error) {
      console.error('Error creating damage scenario:', error);
      throw error;
    }
  },

  /**
   * Update an existing damage scenario
   * @param id - Damage scenario ID
   * @param scenario - Updated damage scenario data
   * @returns Promise with updated damage scenario
   */
  async update(id: string, scenario: DamageScenarioUpdate): Promise<DamageScenario> {
    return apiClient.put<DamageScenario>(`/damage-scenarios/${id}`, scenario);
  },

  /**
   * Delete a damage scenario
   * @param id - Damage scenario ID
   * @returns Promise with success response
   */
  async delete(id: string): Promise<void> {
    return apiClient.delete<void>(`/damage-scenarios/${id}`);
  },

  /**
   * Get propagation suggestions for a component
   * @param componentId - Source component ID
   * @param impacts - Impact types to consider
   * @returns Promise with propagation suggestions
   */
  async getPropagationSuggestions(
    componentId: string,
    impacts: {
      confidentiality?: boolean;
      integrity?: boolean;
      availability?: boolean;
    } = {}
  ): Promise<PropagationSuggestionResponse> {
    try {
      const queryParams = new URLSearchParams();
      queryParams.append('component_id', componentId);
      
      if (impacts.confidentiality !== undefined) {
        queryParams.append('confidentiality_impact', impacts.confidentiality.toString());
      }
      
      if (impacts.integrity !== undefined) {
        queryParams.append('integrity_impact', impacts.integrity.toString());
      }
      
      if (impacts.availability !== undefined) {
        queryParams.append('availability_impact', impacts.availability.toString());
      }
      
      const response = await apiClient.get<PropagationSuggestionResponse>(
        `/damage-scenarios/propagation/suggestions?${queryParams.toString()}`
      );
      
      return response;
    } catch (error) {
      console.error('Error fetching propagation suggestions:', error);
      return { suggestions: [], total: 0 };
    }
  }
};

export default damageScenarioApi;
